File: crates/1bit-mcp-clients/src/protocol.rs
//! JSON-RPC 2.0 wire types for the subset of MCP we speak on the client
//! side: `initialize`, `tools/list`, `tools/call`.
//!
//! We intentionally do not implement the full MCP spec — no prompts,
//! no resources, no sampling — because halo's outbound use case is
//! tool-calling only. If that changes, add methods here and bump
//! `PROTOCOL_VERSION`.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// MCP protocol version we advertise in `initialize`. Matches the
/// `2025-06-18` revision.
pub const PROTOCOL_VERSION: &str = "2025-06-18";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Request<'a> {
    pub jsonrpc: &'a str,
    pub id: u64,
    pub method: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Response {
    #[allow(dead_code)]
    pub jsonrpc: String,
    pub id: Option<u64>,
    #[serde(default)]
    pub result: Option<Value>,
    #[serde(default)]
    pub error: Option<RpcErrorPayload>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RpcErrorPayload {
    pub code: i64,
    pub message: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct Tool {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default, rename = "inputSchema")]
    pub input_schema: Value,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ToolCallResult {
    #[serde(default)]
    pub content: Vec<ContentBlock>,
    #[serde(default, rename = "isError")]
    pub is_error: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ContentBlock {
    Text {
        text: String,
    },
    #[serde(other)]
    Other,
}

impl ContentBlock {
    pub fn as_text(&self) -> Option<&str> {
        match self {
            ContentBlock::Text { text } => Some(text),
            _ => None,
        }
    }
}

pub fn build_request(id: u64, method: &str, params: Option<Value>) -> Value {
    serde_json::json!({
        "jsonrpc": "2.0",
        "id": id,
        "method": method,
        "params": params,
    })
}

pub fn initialize_params(client_name: &str, client_version: &str) -> Value {
    serde_json::json!({
        "protocolVersion": PROTOCOL_VERSION,
        "capabilities": {},
        "clientInfo": { "name": client_name, "version": client_version }
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn protocol_version_is_known_revision() {
        assert_eq!(PROTOCOL_VERSION, "2025-06-18");
    }

    #[test]
    fn build_request_shape_matches_jsonrpc_2() {
        let v = build_request(7, "tools/list", None);
        assert_eq!(v["jsonrpc"], "2.0");
        assert_eq!(v["id"], 7);
        assert_eq!(v["method"], "tools/list");
    }

    #[test]
    fn tool_deserializes_loose_schema() {
        let raw = r#"{ "name":"x","description":"d" }"#;
        let t: Tool = serde_json::from_str(raw).unwrap();
        assert_eq!(t.name, "x");
        assert_eq!(t.description, "d");
        assert!(t.input_schema.is_null());
    }

    #[test]
    fn content_block_text_extracts() {
        let raw = r#"{"type":"text","text":"hello"}"#;
        let b: ContentBlock = serde_json::from_str(raw).unwrap();
        assert_eq!(b.as_text(), Some("hello"));
    }

    #[test]
    fn content_block_unknown_is_other() {
        let raw = r#"{"type":"image","data":"..."}"#;
        let b: ContentBlock = serde_json::from_str(raw).unwrap();
        assert!(b.as_text().is_none());
    }
}

File: cpp/mcp-clients/src/protocol.cpp
#include "onebit/mcp_clients/protocol.hpp"

namespace onebit::mcp_clients {

void to_json(json& j, const Tool& t)
{
    j = {{"name", t.name}, {"description", t.description}};
    if (!t.input_schema.is_null()) {
        j["inputSchema"] = t.input_schema;
    }
}

void from_json(const json& j, Tool& t)
{
    t.name        = j.value("name", std::string{});
    t.description = j.value("description", std::string{});
    if (j.contains("inputSchema")) {
        t.input_schema = j.at("inputSchema");
    } else {
        t.input_schema = nullptr;
    }
}

void to_json(json& j, const ContentBlock& b)
{
    if (b.kind == ContentBlockKind::Text) {
        j = {{"type", "text"}, {"text", b.text}};
    } else {
        j = {{"type", "other"}};
    }
}

void from_json(const json& j, ContentBlock& b)
{
    const std::string type = j.value("type", std::string{});
    if (type == "text") {
        b.kind = ContentBlockKind::Text;
        b.text = j.value("text", std::string{});
    } else {
        b.kind = ContentBlockKind::Other;
        b.text.clear();
    }
}

void to_json(json& j, const ToolCallResult& r)
{
    j = {{"content", r.content}, {"isError", r.is_error}};
}

void from_json(const json& j, ToolCallResult& r)
{
    r.content.clear();
    if (j.contains("content") && j.at("content").is_array()) {
        for (const auto& item : j.at("content")) {
            r.content.push_back(item.get<ContentBlock>());
        }
    }
    r.is_error = j.value("isError", false);
}

json build_request(std::uint64_t id, std::string_view method,
                   const std::optional<json>& params)
{
    json out = {
        {"jsonrpc", "2.0"},
        {"id",      id},
        {"method",  std::string(method)},
        {"params",  params.has_value() ? *params : json(nullptr)},
    };
    return out;
}

json initialize_params(std::string_view client_name, std::string_view client_version)
{
    return {
        {"protocolVersion", std::string(PROTOCOL_VERSION)},
        {"capabilities",    json::object()},
        {"clientInfo",      {
            {"name",    std::string(client_name)},
            {"version", std::string(client_version)},
        }},
    };
}

} // namespace onebit::mcp_clients
